### PR TITLE
[NNX] NNX migration prep (7.5/N): finish MaxEngine inference carve-outs

### PR DESCRIPTION
**Pull request by ecnal-cienet**
## NNX Migration Route Map

1. ✅ NNX scaffolding: `pure_nnx` flag, `init_state_fn`, `TrainStateNNX`, NNX utils. (PR #3427)
2. ✅ NNX sharding utilities. (PR #3470)
3. ✅ NNX end-to-end training loop: model creation, gradient accumulation, checkpointing, dispatch. (PR #3500)
4. ✅ Sharding diagnostics on NNX + post-training bugfixes. (PR #3652)
5. ✅ NNX correctness fixes, feature enablements, and vocab tiling on NNX.
6. ✅ NNX-native DPO.
7. ✅ NNX-native MaxEngine inference (core prefill/generate/insert path). (PR #3821)
7.5. 🔄 **[This PR]** Finish the NNX MaxEngine inference carve-outs.
8. ❌ NNX-native LoRA + GRPO.
9. ❌ NNX-aware QK-Clip + remaining checkpoint utilities.
9.5. ❌ NNX + AQT in MaxEngine + serve-mode reload + gpt3 prefill fix.
10. ❌ Vocab tiling `custom_vjp` for NNX.
11. ❌ Set NNX defaults to `True`.
12. ❌ Delete Linen-specific code paths and NNX compatibility flags.

## Description

PR7 made the core MaxEngine inference path (prefill, generate, insert, `load_params`, `init_decode_state`) NNX-native under `pure_nnx=True`, but deliberately left three serving features raising `NotImplementedError` on the NNX path. This PR promotes all three to NNX-native, reusing the patterns PR7 established (`_nnx_run_model` at the JIT boundary, pure-dict cache threading).

Linen is preserved byte-for-byte: the original `self.model.apply(..., mutable=["cache"])` calls are unchanged, just moved into `else:` branches, and every NNX edit is gated `if config.pure_nnx:`. `pure_nnx=False` stays the default.

Diff: **1 source file + 1 test file**. Stacks on PR7 (`feat/nnx-native-maxengine`).

## What this implements

- **`prefill_multisampling`** (`_prefill_multisampling_jit`): drops the `NotImplementedError`; the `pure_nnx` branch runs the batch=1 prefill through `_nnx_run_model` (`MODEL_MODE_PREFILL`) with a fresh `_nnx_init_cache_dict`. The loop that draws `num_samples` first tokens from the shared logits is unchanged. Mirrors the `_prefill_jit` NNX branch.
- **`prefill_concat`**: same swap; the packed `decoder_positions` / `decoder_segment_ids` thread through `_nnx_run_model` unchanged.
- **`stack_prefill_result_cache=True`**: the hard one. The NNX cache pure-dict differs from Linen, and differs by scan mode:

  | Config | Cache layout | Leaf shape | Handling |
  |---|---|---|---|
  | NNX `scan_layers=True` | one `decoder/layers` node | `[N, ...]` (layer axis present) | no-op — already stacked |
  | NNX `scan_layers=False` | per-layer `decoder/layers/{0,1}` (int keys) | `[...]` (no layer axis) | real stack/unstack |
  | Linen unstacked | flat `decoder/layers_0`, `layers_1` (str keys) | `[...]` | existing Linen helper |

  So for `scan_layers=True` the cache is already in the Linen post-stack shape: `_maybe_stack_prefill_result_cache` / `_maybe_unstack_prefill_result_cache` are no-ops and `prefill_kv_cache_shardings` stays the full per-leaf tree. For `scan_layers=False`, `_maybe_stack` collapses the per-layer subtrees into one `decoder/layers` subtree with a leading layer axis, `_maybe_unstack` splits it back into the int-keyed per-layer dict that `bulk_insert` / `_insert_jit` / `insert_partial` walk, and `_load_params_nnx` prepends a layer axis to each prefix-sharding spec (the NNX analog of the Linen `P(None, *spec)` + `["decoder"]["layers_0"]` reshape).

## Carve-outs

The only remaining NNX MaxEngine carve-outs belong to other PRs in the stack:

| Site | Tracked in |
|------|------------|
| quantization (`quantize_params`, quantized-checkpoint load) | PR9.5 |
| LoRA (`load_single_adapter`) | PR8 |

`prefill_multisampling`, `prefill_concat`, and `stack_prefill_result_cache=True` (both `scan_layers` values) are no longer stubbed.

## Tests

`tests/integration/maxengine_test.py` — 4 new NNX tests + a `_build_linen_params` helper and a shared `_stack_prefill_roundtrip` helper:

- `test_prefill_multisampling_nnx`: NNX result shape matches the Linen path for `num_samples=3`; logits and every cache leaf are finite.
- `test_prefill_concat_nnx`: NNX result shape matches the Linen path for a 2-prompt packed prefill; finite.
- `test_stack_prefill_result_cache_nnx` (`scan_layers=True`) and `test_stack_prefill_result_cache_scan_layers_false_nnx` (`scan_layers=False`): prefill -> insert -> generate round-trip with the flag on; cache leaves carry the layer axis on axis 0; logits stay finite; `next_pos` advances each step.

### feat/nnx-native-maxengine Branch
- Linen:
    - Terminal 1: https://paste.googleplex.com/6581568998080512
    - Terminal 2: https://paste.googleplex.com/4994926726217728
    - xprof: https://xprof.corp.google.com/memory_profile/wanglance-7438713142362131305

- NNX:
    - Terminal 1: https://paste.googleplex.com/4809150952505344
    - Terminal 2: https://paste.googleplex.com/6398572420530176
    - xprof: https://xprof.corp.google.com/memory_profile/wanglance-758614634353888613
 
## Inference Microbenchmark Test
- feat/nnx-native-maxengine Linen vs NNX: https://diff.googleplex.com/#key=zRk5JxAGma8k

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by bvandermoon — COMMENTED** [review]
Was this PR run with the same manual tests as https://github.com/AI-Hypercomputer/maxtext/pull/3821?

**Comment by ecnal-cienet** [comment]
> Was this PR run with the same manual tests as #3821?

Hi @bvandermoon,
Yes, I did the same Maxengine manual test as well as Inference Microbenchmark Test for this PR. I posted the results in PR body.

**Review by bvandermoon — APPROVED** [review]

File: src/maxtext/inference/maxengine/maxengine.py [hunk 1]
```diff
@@ -272,11 +363,90 @@ def load_params(self, *args, params=None, rng: PRNGKeyType | None = None, **kwar
...
+    if self.model.quant is not None:
+      raise NotImplementedError("pure_nnx + quantization not yet supported. Use pure_nnx=False.")
+
+    if params:
+      print("Resharding given NNX params")
+      _, params_abs, _ = nnx.split(self.model, nnx.Param, ...)
+      # self.model is abstract (built via nnx.eval_shape), so its leaves carry logical
+      # axis metadata but no physical .sharding. Resolve logical to physical here so
+      # device_put actually reshards instead of being a no-op.
+      with nn_partitioning.axis_rules(self.config.logical_axis_rules):
```
Comment (NuojCheng) [inline src/maxtext/inference/maxengine/maxengine.py:383]: with jax.set_mesh(self.mesh), nn_partitioning.axis_rules(self.config.logical_axis_rules):

File: src/maxtext/inference/maxengine/maxengine.py [hunk 2]
```diff
@@ -272,11 +363,90 @@ def load_params(self, *args, params=None, rng: PRNGKeyType | None = None, **kwar
...
+        target_shardings = maxtext_utils.get_nnx_named_sharding_with_scan_axis(params_abs, self._mesh)
+      params_state = jax.device_put(params, target_shardings)
+      # We only need a concrete `rest` (RNG vars) for nnx.merge. create_nnx_sharded_model
+      # builds the model with a jitted out_shardings so params are produced already
+      # sharded, avoiding a single-device allocation of the full model (an OOM risk for
+      # large models). self.model is abstract with no .sharding, so pass an explicit one.
+      _, full_abs = nnx.split(self.model)
+      with nn_partitioning.axis_rules(self.config.logical_axis_rules):
```
Comment (NuojCheng) [inline src/maxtext/inference/maxengine/maxengine.py:391]: same above

File: src/maxtext/utils/maxtext_utils.py [hunk 3]
```diff
@@ -1812,6 +1812,30 @@ def init_kv_cache(model, config):
...
+def _nnx_cache_partition_specs(abstract_model, config, mesh):
+  """Per-leaf PartitionSpec tree for the abstract model's nnx.Cache vars.
+
+  Returned as a pure dict so the engine can wrap it in NamedSharding the same
+  way it does for the Linen helpers below.
+  """
+  _, cache_state, _ = nnx.split(abstract_model, nnx.Cache, ...)
+  # get_nnx_named_sharding_with_scan_axis reads logical axis rules from the
+  # active flax partitioning context, so wrap.
+  with nn_partitioning.axis_rules(config.logical_axis_rules):
```
Comment (NuojCheng) [inline src/maxtext/utils/maxtext_utils.py:1836]: in general we recommend with jax.set_mesh(mesh), nn_partition...

**Review by NuojCheng — APPROVED** [review]
LGTM

**Review by cgarciae — APPROVED** [review]
